### DISCUSSION TITLE
How to implement high-level wrappers for classes like `SkSVGPaint`?

### DISCUSSION THREAD

**Discussion by Aiving** [rust-skia/rust-skia#1027] [Q&A] [answered]
I also decided to improve skia-safe's SVG support (in #1025) by adding `set` and `get` methods for all attributes, but ran into the fact that there are classes like `SkSVGPaint`. They don't fit under `RCHandle` either, because they don't have a reference counter (for obvious reasons), but they don't fit under `Handle` either, because they don't have a method for `Drop`. At the same time, the same `SkSVGPaint` is used for the `fill` attribute, so implementing wrapper for it is also important.

![image](https://github.com/user-attachments/assets/efb4d983-465b-44e6-aa51-83d840bf6ed3)

**Comment by pragmatrix** [comment]
> because they don't have a method for Drop.

@Aiving Even though there is no explicit destructor, this type has an implicitly generated implementation, because for example SkSVGColor contains a smartpointer reference. So the my first choice would be to use `Handle`, implement `NativeDrop` and forward this to a C function that invokes the destructor.

**Reply by Aiving** [reply]
> > because they don't have a method for Drop.
> 
> @Aiving Even though there is no explicit destructor, this type has an implicitly generated implementation, because for example SkSVGColor contains a smartpointer reference. So the my first choice would be to use `Handle`, implement `NativeDrop` and forward this to a C function that invokes the destructor.
> 
> 
> 

Will try soon!

**Reply by Aiving** [reply]
> implement `NativeDrop` and forward this to a C function that invokes the destructor.

Are you talking about implementing a function like `C_SkCodec_delete`?

![image](https://github.com/user-attachments/assets/a01c64ac-c901-48c5-ae4e-2e5a043441a7)

**Reply by pragmatrix** [reply]
No, `Handle<>` directly manages the memory of the C++ in Rust, so the destructor must be called in-place (and the constructor, too, using placement new), for example `Handle<SkBitmap>`:

```c++
extern "C" void C_SkBitmap_destruct(SkBitmap* self) {
    self->~SkBitmap();
}
```

BTW: `SkCodec` is wrapped as a `RefHandle<SkCodec>` in Rust.

This is for cases the layout does not work, or if self-referential pointers are in the C++ type, then a `new` and `delete` wrapper function is needed and the memory is managed by C++.

**Reply by Aiving** [reply]
I'm not quite sure what to do. If I try to use `Handle<sb::SkSVGPaint>`, there is a problem with how to return the owned value when creating (`from_native_ptr_mut` returns `*mut Handle<T>`, perhaps I should use `read` as in the screenshot?). There is `RefHandle`, but it requires a `*mut T` reference to create, and `getStroke` returns `*const sb::SkSVGPaint`.

![image](https://github.com/user-attachments/assets/29a38d9b-bd86-4be0-a2e3-d68dfabb6636)

**Reply by pragmatrix** [reply, marked as answer]
For the creation of new Handle<T> values use the following pattern:

Rust:

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-safe/src/gpu/ganesh/vk/vk_backend_surface.rs#L17-L19 

`construct` allocates memory in Rust, then pass the pointer to the in-place new function.

C++

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-bindings/src/vulkan.cpp#L19-L21

And if you receive a const pointer to the native C++ type, use `Handle::from_native_ref` using an `unsafe {&*ptr)` conversion.

Note that this causes undefined behavior if the pointer is null. In these cases you may use

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-safe/src/prelude.rs#L33

to convert a pointer to an Option<*const ptr> and then use `.map` with `Handle::from_native_ref` like so:

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-safe/src/effects/runtime_effect.rs#L322-L324

**Reply by Aiving** [reply]
I wanted to add also functions like `children` and `children_mut` for `SkSVGSVG`, but unfortunately `fChildren` is a `protected` property, so there is no way to create a wrapper function (without a potential UB like `static_cast` to my own type). Also, it seems that working with dynamic arrays/vectors in bindgen is not the nicest thing to do...  Do you have any ideas? Or should I just leave it at that?

![image](https://github.com/user-attachments/assets/e5c1d6e2-681c-43a4-8119-04391d1cc52a)

**Reply by pragmatrix** [reply]
Hmm, weird that access to children is not public. The TODO seems to indicate that this is not finalized yet.

For returning arrays/vectors from C++ to Rust you can use VecSink:

C++:

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-bindings/src/paragraph.cpp#L90-L100

Rust:

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-safe/src/modules/paragraph/font_collection.rs#L138-L144

This is not quite efficient, but a good (safe) start. I am not sure how fChildren is layed out, but if it's a contiguous array of SkSVGNode, you may return a pointer and length, create a slice in Rust and transmute it to `&[RCHandle<SkSVGNode>]` ... but this would expose implementation details. :thinking: I think a better alternative is to add a C++ function that takes an index and returns a pointer to a SkSVGNode and create a custom Rust iterator on top of that.

> Or should I just leave it at that?

I've no idea how the API should be used, but I personally would be rather confused if nodes can be added to a container, but not removed or iterated.

**Reply by pragmatrix** [reply]
> Or should I just leave it at that?

Also fine with me, after all, this is progress and better than before.

**Reply by Aiving** [reply]
@pragmatrix Have you encountered that despite adding `#include "path/to/header.h"`, `bindgen` does not generate types from this `header.h`? I have this situation right now, and `bindgen` only starts generating types if I add the `extern "C"` function, which uses types from `header.h` (which is obviously a bit inconvenient)

**Reply by pragmatrix** [reply]
Yes, this is "by design". Bindgen is configured to create wrappers only for functions that start with `C_` and all transitively referred types. In such situations, I just add a _dummy_ function with pointers to the types that need bindings, like in gl.cpp for example:

https://github.com/rust-skia/rust-skia/blob/dc17b3e9b36b91c342ef27741d28d701d056d5c8/skia-bindings/src/gl.cpp#L20

**Reply by Aiving** [reply]
I'm still working on expanding SVG support and ran into a new question. Skia has a "base" class `SkSVGNode` which is inherited by all elements that can be in `fChildren` of containers (this includes `SkSVGSVG`, `SkSVGG`, etc.). I want to make a wrapper over it, but I don't quite understand how to implement it. Since the number of classes is limited, I was thinking of a variant using enums (something like `SvgNode::attributes(&self) -> SvgNodeAttributes`, `enum SvgNodeAttributes { Circle(SvgCircle), ... }`), but this seems quite complicated to implement. Perhaps there are some other options or an easier way to implement something like this?
I also had to add `paste` as a dependency because writing all the `get` and `set` methods manually is horrible. (probably remove it and replace all the code with what the macro generates when I finish working on extending SVG support)

![image](https://github.com/user-attachments/assets/23b2142a-65ff-4f9c-b6b2-627864f229bf)

**Reply by Aiving** [reply]
There seems to be something wrong with getting the correct list of elements from `fChildren` (random number on the screenshot - the value of the `fTag` field that exceeds its bounds)
I also checked the elements in `fChildren` on the C++ side and everything is fine there... 

![image](https://github.com/user-attachments/assets/21ff110f-3857-4bd7-92de-b86569e31886)

**Reply by pragmatrix** [reply]
@Aiving May be a bindgen layout failure, try running `cargo test -p skia-bindings`. If that succeeds, I'd like to take a look. Open a PR or point me to the branch that reproduces the problem.

**Reply by Aiving** [reply]
> @Aiving May be a bindgen layout failure, try running `cargo test -p skia-bindings`. If that succeeds, I'd like to take a look. Open a PR or point me to the branch that reproduces the problem.

I'm ashamed to admit it, but I've already solved that problem. It turned out that the issue was in what type of elements `fChildren` stores. They are not `SkSVGNode` directly, but they inherit it. Obviously, the size of an element of type `SkSVGCircle` and another element of type `SkSVGPath` will be different, which means you need a different offset to get any of these two elements correctly. However, Rust has no such inheritance and so it uses the `SkSVGNode` size to determine the correct offset for the elements in the slice created by `slice::from_raw_parts`. The solution was to use `sk_sp` on the C++ side, which only stores a reference to the element and therefore always has the same size, whether in C++ or Rust.
